### PR TITLE
Bw 264

### PR DESCRIPTION
**Pull request by mehrlerKinsol**
Adds equal error rate metric and outputs metrics to a JSON file to work with DVC

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -33,11 +33,12 @@ RUN apt-get update && apt-get install -y apt-utils && apt-get -y dist-upgrade &&
 
 RUN ln -s /usr/bin/python${PYTHON_VER} /usr/bin/python
 
-RUN pip install --no-cache-dir torch==1.7.1+cu110 torchvision==0.8.2+cu110 -f https://download.pytorch.org/whl/torch_stable.html
+RUN pip install --no-cache-dir torch==1.9.0+cu111 torchvision==0.10.0+cu111 torchaudio==0.9.0 -f https://download.pytorch.org/whl/torch_stable.html
```
Comment (mehrlerKinsol) [inline Dockerfile:36]: Not sure why these are showing up as changed. Torch 1.9 should be in master already

File: yolo/utils/metrics.py [hunk 2]
```diff
@@ -138,3 +142,16 @@ def compute_ap(recall, precision):
...
+    """ Computes the equal error rate precision and recall from the precision and recall curves.
+    # Arguments
+        recall:    Recall curve (n sized list of recalls along the recall curve, evaluated at equal intervals)
+        precision: Precision curve (n sized list of recalls along the recall curve, evaluated at equal intervals)
+    # Returns
+        The equal error rate precison, recall, and thershold. Returns NaN if equal error rate could not be found (shouldn't happen with an actual classifier)
+    """
+    for i, (precision,recall) in enumerate(zip(precisions,recalls)):
+        if recall >= precision:
+            return recall,precision,i/recalls.shape[0]
```
Comment (kweston) [inline yolo/utils/metrics.py]: This isn't returning the actual threshold. I think you need to pass the thresholds in as a 3rd function argument and return the one that triggers the change.

File: yolo/utils/metrics.py [hunk 3]
```diff
@@ -91,6 +92,10 @@ def ap_per_class(tp, conf, pred_cls, target_cls, plot=False, fname='precision-re
                 ap[ci, j], mpre, mrec = compute_ap(recall[:, j], precision[:, j])
                 if j == 0:
                     py.append(np.interp(px, mrec, mpre))  # precision at mAP@0.5
+            
+            # remap precision and recall curves at iou .5 from objectness to confidence thresholds
+            remapped_precision, remapped_recall = np.interp(np.flip(-px), -conf[i], precision[:, 0]),np.interp(np.flip(-px), -conf[i], recall[:, 0])
```
Comment (kweston) [inline yolo/utils/metrics.py]: Not sure why you are interpolating here, can't you just read in the curves corresponding to the 0.5 IoU threshold directly?

File: test.py [hunk 4]
```diff
@@ -253,10 +253,20 @@ def test(data,
...
         p, r, ap50, ap = p[:, 0], r[:, 0], ap[:, 0], ap.mean(1)  # [P, R, AP@0.5, AP@0.5:0.95]
         mp, mr, map50, map = p.mean(), r.mean(), ap50.mean(), ap.mean()
         nt = np.bincount(stats[3].astype(np.int64), minlength=nc)  # number of targets per class
+        
+        #output to json for dvc metrics
+        data = {"equal_error_rate":{"precision":eer_precision.mean(),
+                                    "recall":eer_recall.mean(),
+                                    "threshold":threshold.mean()},
+                "average_precision":map50}
+        filename = 'metrics/dev/metrics.json'
```
Comment (kweston) [inline test.py]: I think the file Path object should be passed in as an optional CLI arg. Only write the JSON file if the filename is present

File: test.py [hunk 5]
```diff
@@ -253,10 +253,20 @@ def test(data,
...
+        os.makedirs(os.path.dirname(filename), exist_ok=True)
```
Comment (kweston) [inline test.py]: I prefer to use the pathlib.Path API where possible.
`filename.parent.mkdir(exist_ok=True, parents=True)`

File: yolo/utils/metrics.py [hunk 6]
```diff
@@ -91,6 +92,10 @@ def ap_per_class(tp, conf, pred_cls, target_cls, plot=False, fname='precision-re
...
+            eer_precision[ci],eer_recall[ci],threshold[ci] = compute_eer_pr(remapped_recall,remapped_precision)
```
Comment (kweston) [inline yolo/utils/metrics.py]: I think this should be only called if a `computer_eer` function arg is present.

**Review by kweston — CHANGES_REQUESTED** [review]

File: yolo/utils/metrics.py (hunk 2 above)
Comment (mehrlerKinsol) [reply yolo/utils/metrics.py]: The thresholds in the conf array don't quite map to the right threshold so some kind of mapping is needed. I found a bit cleaner way to do it though

File: yolo/utils/metrics.py (hunk 6 above)
Comment (mehrlerKinsol) [reply yolo/utils/metrics.py]: If this argument is false and a filename for metrics is given what should it output? Just 0s?

Comment (kweston) [reply yolo/utils/metrics.py]: 0s or NaNs I'm fine with either

**Review by kweston — APPROVED** [review]
Remember to squash commits before merging. Maybe we should hold off on merging this until the torch1.9 issues are sorted out.

Also, we usually format our PRs as
BW-264 <short description of ticket>
- <first change>
- <second change>
- ...

**Pull request merged**